pyfr/integrators/dual/phys/steppers.py

```diff
@@ -1,25 +1,125 @@
+        psnregs = self.pseudointegrator.pseudo_stepper_nregs
+        snregs = self.pseudointegrator.stepper_nregs
 
-class DualBDF2Stepper(BaseDualStepper):
+        # Rotate the source registers to the right by one
+        self.pseudointegrator._regidx[psnregs:psnregs + snregs] = (
+            self.pseudointegrator._stepper_regidx[-1:]
+            + self.pseudointegrator._stepper_regidx[:-1]
+        )
+
+        # Copy the current soln into the first source register
+        self.pseudointegrator._add(
+            0, self.pseudointegrator._stepper_regidx[0],
+            1, self.pseudointegrator._idxcurr
+        )
+
+    @property
+    def stepper_coeffs(self):
+        return [1] + [sc/self._dt for sc in self.stepper_static_coeffs]
+
+
+class DualBDF2Stepper(BaseBDFStepper):
     stepper_name = 'bdf2'
     stepper_order = 2
-    stepper_coeffs = [-1.5, 2.0, -0.5]
+    stepper_static_coeffs = [-1.5, 2.0, -0.5]
 
 
-class DualBDF3Stepper(BaseDualStepper):
+class DualBDF3Stepper(BaseBDFStepper):
     stepper_name = 'bdf3'
     stepper_order = 3
-    stepper_coeffs = [-11.0/6.0, 3.0, -1.5, 1.0/3.0]
+    stepper_static_coeffs = [-11.0/6.0, 3.0, -1.5, 1.0/3.0]
 
 
-class DualBackwardEulerStepper(BaseDualStepper):
+class DualBackwardEulerStepper(BaseBDFStepper):
     stepper_name = 'backward-euler'
     stepper_order = 1
-    stepper_coeffs = [-1.0, 1.0]
+    stepper_static_coeffs = [-1.0, 1.0]
+
+
+class BaseDIRKStepper(BaseDualStepper):
+    @property
+    def stepper_nregs(self):
+        return 1
+
+    @property
+    def stage_nregs(self):
+        return self._nstages
+
+    def _finalize_stage(self, tcurr):
+        # Finalize DIRK
+        if not self.fsal:
+            # Get the new soln
+            self.pseudointegrator._add(
+                0, self.pseudointegrator._idxcurr,
+                1, self.pseudointegrator._stepper_regidx[0],
+                *chain(*zip([bred*self._dt for bred in self.b],
+                            self.pseudointegrator._stage_regidx)),
+                subdims=self.pseudointegrator._subdims
+            )
+
+        # Copy the new solution into stepper register
+        self.pseudointegrator._add(
+            0, self.pseudointegrator._stepper_regidx[0],
+            1, self.pseudointegrator._idxcurr
+        )
+
+    @property
+    def stepper_coeffs(self):
+        csn = self.currstg
+        return [self.a[csn][csn]] \
+            + [c/self._dt for c in [-1.0, 1.0]] \
+            + self.a[csn][:csn]
+
+
+class ESDIRK3Stepper(BaseDIRKStepper):
+    stepper_name = 'esdirk3'
+    _nstages = 4
+    fsal = True
+
+    a = [[], [], [], []]
+    a[0] = [0]
+    a[1] = [1767732205903/4055673282236, 1767732205903/4055673282236]
+    a[2] = [2746238789719/10658868560708, -640167445237/6845629431997,
+            1767732205903/4055673282236]
+    a[3] = [1471266399579/7840856788654, -4482444167858/7529755066697,
+            11266239266428/11593286722821, 1767732205903/4055673282236]
+
+    b = a[-1]
```

Comment:
For FSAL make this automatic (have the base constructor set b).

Reply:
I don't recognise this scheme. I pulled this branch or your a while back and have had good results from two other methods that I think you consider. See eq 229 and eq 232 here https://ntrs.nasa.gov/api/citations/20160005923/downloads/20160005923.pdf
Or my version of your branch here https://ntrs.nasa.gov/api/citations/20160005923/downloads/20160005923.pdf

Reply:
I have used this scheme (ESDIRK, third order accurate, see eq 224) in the past and I think this is a good one.

Reply:
I took the one I used from here.
Bijl, H., Carpenter, M., Vatsa, V., & Kennedy, C. (2002). Implicit Time Integration Schemes for the Unsteady Compressible Navier–Stokes Equations: Laminar Flow. Journal of Computational Physics, 179, 313-329.
I'm happy to add all these schemes you suggest, thanks.